cuqipy_fenics/testproblem.py

```diff
@@ -147,3 +155,275 @@ def form(m,u,p):
+                is_par=True,
+                geometry=G_domain)
+
+        elif callable(exactSolution):
+            exactSolution_expr = to_dolfin_expression(exactSolution)
+            exactSolution_func = dl.interpolate(exactSolution_expr, V)
+            exactSolution = cuqi.array.CUQIarray(
+                exactSolution_func,
+                is_par=False,
+                geometry=G_domain)
+
+        else:
+            raise ValueError(
+                'exactSolution should be a numpy array, a function or None.')
+        print(exactSolution.__class__)
+
+        # Create the exact data
+        exact_data = A(exactSolution)
+        if not isinstance(exact_data, cuqi.array.CUQIarray):
+            exact_data = cuqi.array.CUQIarray(
+                exact_data, is_par=True, geometry=G_range)
+
+        # Create the data distribution and the noisy data
+        noise = np.random.randn(len(exact_data))
+        if relative_noise_std is None:
+            relative_noise_std = 0.01
+        noise_std = relative_noise_std * \
+            np.linalg.norm(exact_data)/np.linalg.norm(noise)
+        noise = noise_std*noise
+        data = exact_data + noise
+
+        y = Gaussian(mean=A(prior), cov=noise_std**2, geometry=G_range)
+
+        # Create the Bayesian problem
+        super().__init__(y, prior, y=data)
+
+        # Store exact values and information
+        self.exactSolution = exactSolution
+        self.exactData = exact_data
+        self.infoString = f"Noise type: Additive i.i.d. noise with standard deviation: {noise_std} and relative noise standard deviation: {relative_noise_std}."
+
+    def _create_boundaries_subdomains(self):
+        """
+        Create subdomains for the boundary conditions.
+        """
+        class Left(dl.SubDomain):
+            def inside(self, x, on_boundary):
+                return on_boundary and x[0] < dl.DOLFIN_EPS
+
+        class Bottom(dl.SubDomain):
+            def inside(self, x, on_boundary):
+                return on_boundary and x[1] < dl.DOLFIN_EPS
+
+        class Right(dl.SubDomain):
+            def inside(self, x, on_boundary):
+                return on_boundary and x[0] > 1.0 - dl.DOLFIN_EPS
+
+        class Top(dl.SubDomain):
+            def inside(self, x, on_boundary):
+                return on_boundary and x[1] > 1.0 - dl.DOLFIN_EPS
+
+        return [Left(), Bottom(), Right(), Top()]
+
+    def _set_up_dirichlet_bcs(self, V, bc_types, bc_values, subdomains):
+        """
+        Set up Dirichlet boundary conditions for the Poisson PDE problem defined
+        on the unit square mesh, where V is the function space.
+        """
+        dirichlet_bcs = []
+
+        for i, bc in enumerate(bc_types):
+            if bc.lower() == 'dirichlet':
+                dirichlet_bcs.append(dl.DirichletBC(
+                    V, bc_values[i], subdomains[i]))
+
+        return dirichlet_bcs
+
+    def _set_up_adjoint_dirichlet_bcs(self, V, bc_types, subdomains):
```

Comment:
Consider describing somewhere in the docstring what the adjoint bcs are and how they are used.